src/server/twemcache-rs/src/worker/multi.rs

```diff
@@ -0,0 +1,324 @@
+    pub fn run(&mut self) {
+        let mut events = Events::with_capacity(self.config.worker().nevent());
+        let timeout = Some(std::time::Duration::from_millis(
+            self.config.worker().timeout() as u64,
+        ));
+
+        loop {
+            increment_counter!(&Stat::WorkerEventLoop);
+
+            // get events with timeout
+            if self.poll.poll(&mut events, timeout).is_err() {
+                error!("Error polling");
+            }
+
+            increment_counter_by!(
+                &Stat::WorkerEventTotal,
+                events.iter().count().try_into().unwrap(),
+            );
+
+            // process all events
+            for event in events.iter() {
+                self.handle_event(&event);
+            }
+
+            // if we sent any messages to the storage thread, we need to wake it
+            if self.wake_storage && self.storage_queue.wake().is_ok() {
+                self.wake_storage = false;
+            }
+
+            // poll queue to receive new sessions
+            self.handle_new_sessions();
+
+            // poll queue to receive new messages
+            #[allow(clippy::never_loop)]
+            while let Ok(message) = self.message_receiver.try_recv() {
+                match message {
+                    Message::Shutdown => {
+                        return;
+                    }
+                }
+            }
+        }
+    }
+
+    fn handle_event(&mut self, event: &Event) {
+        let token = event.token();
+
+        if token.0 == usize::MAX {
+            self.handle_storage_queue();
+            return;
+        }
+
+        // event for existing session
+        trace!("got event for session: {}", token.0);
+
+        // handle error events first
+        if event.is_error() {
+            increment_counter!(&Stat::WorkerEventError);
+            self.handle_error(token);
+        }
+
+        // handle write events before read events to reduce write buffer
+        // growth if there is also a readable event
+        if event.is_writable() {
+            increment_counter!(&Stat::WorkerEventWrite);
+            self.do_write(token);
+        }
+
+        // read events are handled last
+        if event.is_readable() {
+            increment_counter!(&Stat::WorkerEventRead);
+            let _ = self.do_read(token);
+        }
+
+        if let Some(session) = self.sessions.get_mut(token.0) {
+            trace!(
+                "{} bytes pending in rx buffer for session: {}",
+                session.read_pending(),
+                token.0
+            );
+            trace!(
+                "{} bytes pending in tx buffer for session: {}",
+                session.write_pending(),
+                token.0
+            )
+        }
+    }
+
+    fn handle_session_read(
+        session: &mut Session,
+        poll: &Poll,
+        storage_queue: &mut StorageQueue,
+    ) -> bool {
+        match MemcacheParser::parse(&mut session.read_buffer) {
```

Comment:
I think it's debatable whether this handler should be in worker or `RequestProcessor`. Current location prevents reusing workers for different cache types (e.g. Redis equivalent)